### ISSUE TITLE
Multipart uploads cause an infinite loop if passed sliced data

### ISSUE THREAD

**Issue by harlanhaskins** [awslabs/aws-sdk-swift#1824]
### Describe the bug

`Data` is a self-slicing type, so you can provide `someData[0..<size]` to an UploadPart. However, if you do that, `StreamableHttpBody` gets into a state where it loops forever trying to read 0 bytes.

### Expected Behavior

I would expect self-sliced data to be supported, since that prevents the need for spurious copies during uploading.

### Current Behavior

I tracked it down to this code in `StreamableHttpBody`:

```swift
logger.debug("read \(toRead) bytes from data")
if toRead == 0 && data.endIndex == position {
    return nil
}
```

In this case, `data.endIndex` refers to the end index of the _full_ data object, not just the slice. So this check never succeeds, and we continue reading 0 bytes indefinitely.

### Reproduction Steps

This is the code for an actor that uploads a file in chunks by slicing data.

```swift
public actor FileUploadService {
    public static let shared = FileUploadService()
    private let s3: S3Client

    public init() {
        let config = try! S3Client.S3ClientConfiguration(region: "us-east-1")
        self.s3 = S3Client(config: config)
    }

    public func uploadFile(at url: URL) async throws {
        let id = UUID()
        let input = CreateMultipartUploadInput(acl: .publicRead, bucket: "some-bucket", key: "\(id.uuidString.lowercased())/\(url.lastPathComponent)")
        let upload = try await s3.createMultipartUpload(input: input)
        let data = try Data(contentsOf: url, options: .alwaysMapped)

        var completedParts = [S3ClientTypes.CompletedPart]()

        // Iterate over 1MB chunks
        let chunks = data.chunks(ofSize: 1_000_000)
        for (offset, part) in chunks.enumerated() {
            let partNumber = offset + 1
            Logger.upload.log("Uploading part \(partNumber) of \(chunks.count)...")
            do {
                let partInput = UploadPartInput(body: .data(part), bucket: upload.bucket, contentLength: part.count, key: upload.key, partNumber: partNumber, uploadId: upload.uploadId)
                let partResult = try await s3.uploadPart(input: partInput)
                let completedPart = S3ClientTypes.CompletedPart(eTag: partResult.eTag, partNumber: partNumber)
                completedParts.append(completedPart)
                Logger.upload.log("Finished uploading part \(partNumber) of \(chunks.count)...")
            } catch {
                Logger.upload.error("upload part failed: \(error)")
                let abort = AbortMultipartUploadInput(bucket: upload.bucket, key: upload.key, uploadId: upload.uploadId)
                let _ = try await s3.abortMultipartUpload(input: abort)
                break
            }
        }
        let parts = S3ClientTypes.CompletedMultipartUpload(parts: completedParts)
        let completeInput = CompleteMultipartUploadInput(bucket: upload.bucket, key: upload.key, multipartUpload: parts, uploadId: upload.uploadId)
        let completedUpload = try await s3.completeMultipartUpload(input: completeInput)
        print(completedUpload.location!)
    }
}

// MARK: Helper for chunking data into slices

/// Iterates over fixed-sized consecutive subsequences of a given
/// collection. The final chunk may be smaller than the provided chunk size.
///
/// - Example:
///
/// ```
/// (0..<100).chunks(ofSize: 17)
///
/// // 0..<17
/// // 17..<34
/// // 34..<51
/// // 51..<68
/// // 68..<85
/// // 85..<100
/// ```
///
/// If the collection being chunked is empty, the chunks collection will
/// also be empty.
public struct Chunks<Data: Collection>: RandomAccessCollection {
    public typealias Element = Data.SubSequence

    var chunkSize: Int
    var collection: Data

    public init(_ collection: Data, chunkSize: Int) {
        precondition(chunkSize > 0)
        self.chunkSize = chunkSize
        self.collection = collection
    }

    public subscript(index: Int) -> Element {
        let subrangeStart = collection.index(
            collection.startIndex,
            offsetBy: index * chunkSize,
            limitedBy: collection.endIndex
        )!
        let subrangeEnd =
            collection.index(
                subrangeStart,
                offsetBy: chunkSize,
                limitedBy: collection.endIndex
            ) ?? collection.endIndex
        return collection[subrangeStart..<subrangeEnd]
    }

    public var startIndex: Int { 0 }
    public var endIndex: Int {
        let ratio = Double(collection.count) / Double(chunkSize)
        return Int(ratio.rounded(.up))
    }
    public func index(after i: Int) -> Int { i + 1 }
}

extension Collection {
    public func chunks(ofSize chunkSize: Int) -> Chunks<Self> {
        Chunks(self, chunkSize: chunkSize)
    }
}
```

You can trigger the bug by uploading anything over 1MB.

### Possible Solution

I think by checking `data.count` instead of `data.endIndex`, this will be fully resolved.

### AWS SWIFT SDK version used

1.0.40

### Compiler and Version used

Xcode 16.1 (16B40), Apple Swift version 6.0.2 (swiftlang-6.0.2.1.2 clang-1600.0.26.4), swift-tools-version: 6.0

### Operating System and version

macOS 15.0 (24A8332)

**Comment by jbelkins** [comment]
@harlanhaskins Thank you for the report, and especially for the detailed reproduction instructions.  I didn't get the exact failure as you - I got a fatalError for trying to create an invalid range instead of a hung transfer - but there is definitely a problem when the startIndex isn't zero.

We'll get a fix together soon and update here once it ships.

**Comment by jbelkins** [comment]
@harlanhaskins A fix for this issue shipped in [AWS SDK for Swift 1.0.43](https://github.com/awslabs/aws-sdk-swift/releases/tag/1.0.43) today.

Please update, retest, and if your issue is resolved, please close this ticket.

**Comment by harlanhaskins** [comment]
Confirmed; I removed my workaround and this now behaves as expected with 1.0.43. Thank you!
